[Merged by Bors] - feat(MvPolynomial/WeightedHomogenous): relate `weightedTotalDegree` to `degrees` and `degreeOf`

LiamSchilling opened the pull request:
`weightedTotalDegree` is the most general notion of multivariate degree in mathlib. These theorems show that it specializes to `degrees` and `degreeOf` with specialized weights.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

LiamSchilling committed:
feat(MvPolynomial/WeightedHomogenous): relate `weightedTotalDegree` to `degrees`, `totalDegree`, and `degreeOf`

LiamSchilling committed:
refactor(MvPolynomial/WeightedHomogenous): move theorems relating `weightedTotalDegree` to `degrees`, `totalDegree`, and `degreeOf`

LiamSchilling committed:
fix(MvPolynomial/WeightedHomogenous): fix style and scope issues in theorems relating `weightedTotalDegree` to `degrees`, `totalDegree`, and `degreeOf`

Replace `DecidableEq` inputs with `open Classical`. Change nonterminating `simp` to `simp only`.

LiamSchilling committed:
style(MvPolynomial/WeightedHomogenous): rename `s` for a monomial to `d`

LiamSchilling committed:
feat(MvPolynomial/WeightedHomogenous): declare theorems establishing greedy upper bounds on `weightedTotalDegree`

LiamSchilling committed:
style(MvPolynomial/WeightedHomogenous): fix typo  in declaration name

LiamSchilling committed:
fix(MvPolynomial/WeightedHomogenous): fix typo in declaration type

LiamSchilling committed:
fix(MvPolynomial/WeightedHomogenous): reset to last stable version for a PR, progress will be moved to other branch

chrisflav reviewed (commented):
Thanks for your PR! I left some comments below, feel free to request my review again when you have addressed the comments.

LiamSchilling committed:
feat: address comments for PR

LiamSchilling committed:
fix: remove now unnecessary import

LiamSchilling commented:
Thank you for the comments @chrisflav. To align with the previous definition `weightedTotalDegree_one` you mentioned, I renamed and moved my defintions to the same file.

LiamSchilling committed:
Merge branch 'master' into weighted-total-degree

LiamSchilling committed:
Merge branch 'leanprover-community:master' into weighted-total-degree

LiamSchilling committed:
Update Mathlib/RingTheory/MvPolynomial/Homogeneous.lean

Co-authored-by: Vlad Tsyrklevich <vlad902@gmail.com>

chrisflav reviewed (approved):
Thanks! Modulo the comment, LGTM. Sorry for not responding earlier to this: I did not notice this was awaiting review again, because it was still tagged as `awaiting-author`. For the future feel free to remove the `awaiting-author` label as soon as you have responded to the review comments. Otherwise your PR does not appear on the queue again and many reviewers, including me, only regularly check the PRs on the queue.

LiamSchilling committed:
Update Homogeneous.lean

Co-authored-by: Christian Merten <christian@merten.dev>

chrisflav reviewed (approved):
Thanks!
maintainer merge

LiamSchilling committed:
Update Mathlib/RingTheory/MvPolynomial/Homogeneous.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

LiamSchilling committed:
Update Mathlib/RingTheory/MvPolynomial/Homogeneous.lean

Co-authored-by: Johan Commelin <johan@commelin.net>
